tests/virt_autotest/sriov_pci_passthrough.pm

```diff
@@ -50,88 +45,97 @@ sub run_test {
+    my %vf_vm_2 = plugin_device($guest, $host_vf_2{id});
   
     #check the networking of the plugged interface
     my $gateway = script_output "ip r s | grep 'default via' | cut -d ' ' -f3";
     set_var("SRIOV_PCI_PASSSHTROUGH", 1);   #to differenciate virtual network tests
-    test_network_interface($guest, gate => $gateway, mac => $vf_vm{mac}, net => 'br123');   #use br123 as ssh connection
+    test_network_interface($guest, gate => $gateway, mac => $vf_vm_1{mac}, net => 'br123');   #use br123 as ssh connection
+    test_network_interface($guest, gate => $gateway, mac => $vf_vm_2{mac}, net => 'br123');
   
-    #reboot
+    #reboot the guest
     record_info("VM rebooting", "$guest");
-    script_run "ssh root\@$guest 'reboot'";
-    my $time = 0;
-    while($time < 120 ) {
-      last if ! script_run "ssh root\@$guest 'echo Hello!'";
-      $time += 10;
-      sleep 10;
-    }
+    assert_script_run "ssh root\@$guest 'reboot'";
+    script_retry("nmap $_ -PN -p ssh | grep open", delay => 10, retry => 18,  die => 1);
   
     #check again the networking inside vm
-    test_network_interface($guest, gate => $gateway, mac => $vf_vm{mac}, net => 'br123');
-    set_var("SRIOV_PCI_PASSSHTROUGH", 0);   #turn off to avoid affecting other tests
+    test_network_interface($guest, gate => $gateway, mac => $vf_vm_1{mac}, net => 'br123');
+    test_network_interface($guest, gate => $gateway, mac => $vf_vm_2{mac}, net => 'br123');
   
-    #detach vf from vm via virsh
-    record_info("Unplug VF from vm", "vm=$guest \nmac=$vf_vm{mac}"); 
-    if (unplug_vf_from_vm($guest, $vf_vm{mac})) {
-      $self->{test_results}->{$guest}->{"Error: fail to detach SR-IOV vf network device from $guest!"}->{status} = 'FAILED'; 
+    #detach one vf from vm via virsh
+    record_info("Unplug VFs from vm", "vm=$guest \n \nvf=$host_vf_1{id} \nmac=$vf_vm_1{mac}"); 
+    if (unplug_vf_from_vm($guest, $vf_vm_1{mac})) {
+      $self->{test_results}->{$guest}->{"Error: fail to detach the first SR-IOV Ethernet device from $guest!"}->{status} = 'FAILED'; 
+    }
+
+    #check again the remaining vf inside vm
+    test_network_interface($guest, gate => $gateway, mac => $vf_vm_2{mac}, net => 'br123');
+    set_var("SRIOV_PCI_PASSSHTROUGH", 0);   #turn off to avoid affecting other tests
+
+    #detach the remaining vf from vm via virsh
+    record_info("Unplug VFs from vm", "vm=$guest \n \nvf=$host_vf_2{id} \nmac=$vf_vm_2{mac}");
+    if (unplug_vf_from_vm($guest, $vf_vm_2{mac})) {
+      $self->{test_results}->{$guest}->{"Error: fail to detach the first SR-IOV Ethernet device from $guest!"}->{status} = 'FAILED';       
     }
   
-    #reattach vf to host
-    assert_script_run("virsh nodedev-reattach $host_vf{id}", 60);
+    #reattach vfs to host
+    record_info("Reattach VF to host", "vm=$guest \nvf=$host_vf_1{id} \nvf=$host_vf_2{id}");
+    assert_script_run("virsh nodedev-reattach $host_vf_1{id}", 60);
+    assert_script_run("virsh nodedev-reattach $host_vf_2{id}", 60);
```

Comment:
A suggestion for better handling multiple VFs can be storing all VFs info in arrays and use loop array to avoid duplicating lines, eg this host_vf can look like [{property1=>xx, .., propertyn=>xx}, {property1=>xx, .., propertyn=>xx}, ... ]. Or if even better, you can think about aggregating all variables relevant to VFs to be one hash, including host_vf, vf_vm, and so on. This will look much cleaner.

But since this test only introduces 2 VFs, it is still acceptable to have code as you committed. So you can decide whether to adopt, I am fine here.

Reply:
Did not notice the host vf just match the vf inside vm when the count is increased from one to two. Working on it, changing all functions to passing hash ref as well.